Game logic

men32z opened the pull request.

jstiven01 committed:
Designing Class Game

men32z committed:
show board

jstiven01 committed:
Adding move_game method

men32z committed:
status game

jstiven01 committed:
Adding reset_game method

jstiven01 committed:
Updating user_interface with move_game

men32z committed:
game logic finished

bolah2009 commented on bin/design:
```diff
@@ -0,0 +1,13 @@
+    -status_game (functionality) x
+    -show_board x
+    -reset-game
+    -run
```
Let's delete or keep this file in `.gitignore`

We could actually add this to the readme file to explain the structure of the project or/and this PR description. :thinking: 

> Files not related to the project should not be committed and added to .gitignore.

bolah2009 commented on bin/game.rb:
```diff
@@ -0,0 +1,73 @@
+class Game
```
Let's move this file to the `lib` directory.

> `lib` - for the game logic

bolah2009 commented on bin/game.rb:
```diff
@@ -0,0 +1,73 @@
+  end
+
+  def show_board
+    n_s = [1, 4, 7]
```
[OPTIONAL]
Could we give `n_s` a more descriptive name? :thinking:  It's would make the code easier to read.

bolah2009 commented on bin/main.rb:
```diff
@@ -0,0 +1,45 @@
+    else
+      puts "Next turn"
+    end
+    #Gem.win_platform? ? (system "cls") : (system "clear")
```
#### `REQUIRED`

Consider deleting commented out codes. If you need to review a previous version of your code, use git history.

bolah2009 commented on bin/main.rb:
```diff
@@ -0,0 +1,45 @@
+  while game.total_turns > 0
+    puts "#{game.user[game.current_turn.downcase.to_sym]} pick a number to set the turn '#{game.current_turn}' "
+    position = gets.chomp
+    #if move is valid
```
#### `REQUIRED`

Consider deleting commented out codes. If you need to review a previous version of your code, use git history.

bolah2009 commented on bin/main.rb:
```diff
@@ -0,0 +1,45 @@
+puts "What is your name? (player X)"
+game.user[:x] = gets.chomp
+puts "What is your name? (player O)"
+game.user[:o] = gets.chomp
```
Whenever you find yourself repeating code, know that there's always a way to do it without repeating the code. Keep it DRY.

An example is having a method that asks for names. :thinking: E.g.
```ruby
def get_name(player_name)
  puts  "What is your name? #{player_name}"
  name = gets.chomp
  name = player_name if name.length < 2
 name
end
```
This could be called directly to the `game.user[:x]` and `game.user[:o]` property like:
```ruby
game.user[:x] = get_name("player X")
game.user[:o] = get_name("player O")
```
The advantage of this is that we could change how we get the player's name anytime, like for example there's a validation that sets a player name to a default name if the player inputs a name less than 2 characters. Also check the illutration below on the current effect of the game.

This is what happens when I enter no name. :point_down: 
```sh
Hello welcome to the Tic Tac Toe Game
What is your name? (player X)

What is your name? (player O)

--------------------------------------
This our board
 1 | 2 | 3 
---|---|---
 4 | 5 | 6 
---|---|---
 7 | 8 | 9 
 pick a number to set the turn 'X' 
2
--------------------------------------
Next turn
--------------------------------------
   | X |   
---|---|---
   |   |   
---|---|---
   |   |   
 pick a number to set the turn 'O'
```

Let's compare with this when a valid name is entered. :point_down: 

```sh
Hello welcome to the Tic Tac Toe Game
What is your name? (player X)
Bola
What is your name? (player O)
Ahmed
--------------------------------------
This our board
 1 | 2 | 3 
---|---|---
 4 | 5 | 6 
---|---|---
 7 | 8 | 9 
Bola pick a number to set the turn 'X' 
4
--------------------------------------
Next turn
--------------------------------------
   |   |   
---|---|---
 X |   |   
---|---|---
   |   |   
Ahmed pick a number to set the turn 'O' 
```

> Read about [DRY rule](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself) and use it.

bolah2009 reviewed (changes requested):
## Awesome progress team :clap: 

Let's check the comment below, also note these points from [the project specification on pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/59565018/)
1. Think of how you can encapsulate different parts of the game logic in classes. **Do not create one class for everything**.
2. Read about the DRY rule and use it.
3. Remember about the single responsibility rule. Always.
4. Move class to `lib` directory
5. Fix linter warnings/errors: currently, 81 offenses are detected by `rubocop`. 

:bulb: Use `rubocop` command to check for linter errors locally before committing and/or pushing commits to avoid unnecessary multiple commit messages
:bulb: Use `rubocop -a` to autofix common lint errors
:bulb: If you don't have `rubocop` installed use `gem install rubocop` to install it locally on your machine.

#### :recycle:  Make these changes and submit another code review request. :+1: 

#### Feel free to reach me on slack if you having issues implementing these features. :+1: 

Good luck with your next submission. :star2: 

Cheers
[Bola](https://bolabuari.com/)

###### PS: Note that comments tagged as `OPTIONAL` are not part of the minimum requirements to approve this PR

men32z committed:
first changes

- design file removed.
- class game added but not finished.
- some changes in main logic.

men32z committed:
some errors fixed

- n_s variable changed to a more descriptive name  numbers_with_vertical_sides
- previous comments of code deleted

jstiven01 committed:
Fixing repetitive code
-Creating method to get user's name

jstiven01 committed:
Adding main.rb with rubocop suggestions

maelfosso commented on lib/user.rb:
```diff
@@ -0,0 +1,15 @@
+
+  def set_name
+    puts "What is your name? #{@turn}"
+    @name = gets.chomp
```
No use of `puts`, `get` or `print` in the `lib/` folder

maelfosso commented on bin/main.rb:
```diff
@@ -0,0 +1,46 @@
+usery = User.new('O')
+usery.set_name
+puts '--------------------------------------'
+puts 'This our board'
```
Here, ask the user's name and read here too. Then, use them to instantiate the `User`

maelfosso commented on lib/user.rb:
```diff
@@ -0,0 +1,15 @@
+    @name = gets.chomp
+    @name = 'Player ' + @turn if @name.length < 2
+  end
+end
```
Change to
```ruby
def initialize(name, turn)
   @name = name
   @turn = turn
end
```

Read the need variable outside

maelfosso commented on lib/game.rb:
```diff
@@ -0,0 +1,74 @@
+
+  public
+
+  def move_game(position) # rubocop:disable Metrics/CyclomaticComplexity, Metrics/PerceivedComplexity
```
Remove the comment, please.

maelfosso commented on lib/game.rb:
```diff
@@ -0,0 +1,74 @@
+
+    @total_turns = 0 if winner
+    return "The champion is #{@user[winner.downcase.to_sym].name} '#{winner}'!!" if winner
+  end
```
**`OPTIONAL`**
Don't return a string value. Instead, return a value and use that value to determine the message to print to the user

For example, you could write it like this
```ruby
def move_game(position)
    return :INVALID_POSITION_OCCUPIED unless @board[position - 1].nil?
    return :INVALID_POSITION_OUT_BOUNDARIES unless position >= 1 && position <= 9

    @board[position - 1] = @turn_x ? 'X' : 'O'
    @turn_x = !@turn_x
    @total_turns -= 1
    winner = winner(position - 1)

    return :DRAW if !winner && total_turns < 1

    return winner unless winner

    @total_turns = 0 if winner
    return :CHAMPION if winner
  end
```

Try to do something like this, please.

maelfosso reviewed (changes requested):
Hello @men32z @jstiven01 

It's working :+1:  !!! I manage to play :smile: 
Good job :clap: 

However, as you know make the game works it's not the only thing that matter. There are a few rules that your code should respect. Please, read the comments below.

Kindly, make the changes and submit a new code review request.
Thanks.

jstiven01 committed:
Adding asking names method (code review)

men32z committed:
fixing stickler errors

Torres-ssf commented:
### Pull Request Under Review By Sérgio (TSE)

Torres-ssf reviewed (approved):
### Great job team :clap: 

Nice work implementing the suggested changes, the game works fine and the code is good. :tada: 

**I'm approving this project :+1:**

Good luck with the next milestone.
Happy coding :computer:

men32z committed:
Merge branch 'development' into game_logic
